File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/AgentChatStore.swift
import Foundation
import Observation
import WagerproofModels
import WagerproofServices

/// Per-agent chat thread state. Mirrors RN `useAgentChat` hook (lives in
/// `services/agentChatService.ts` + screen-local state inside AgentChatRoom).
///
/// Optimistic update flow:
///   1. User taps Send → we append a synthetic `AgentChatMessage` to
///      `messages` with a temp id and `role = .user`.
///   2. Service insert fires → on success we swap the temp message for the
///      persisted one (canonical id + timestamp). On failure we strip it and
///      surface the error.
///   3. We then call `requestAssistantReply` which spins up the assistant's
///      response and persists it server-side. The reply lands in `messages`
///      on success.
@Observable
@MainActor
public final class AgentChatStore {
    public enum LoadState: Equatable, Sendable {
        case idle
        case loading
        case loaded
        case failed(String)
    }

    public private(set) var agentId: String
    public private(set) var userId: String?

    public private(set) var messages: [AgentChatMessage] = []
    public private(set) var loadState: LoadState = .idle
    public private(set) var isAssistantTyping: Bool = false
    public private(set) var lastError: String?

    public var draft: String = ""

    public init(agentId: String, userId: String? = nil) {
        self.agentId = agentId
        self.userId = userId
    }

    public func bind(userId: String?) {
        if userId == self.userId { return }
        self.userId = userId
        self.messages = []
        self.loadState = .idle
    }

    /// Load the full thread. Mirrors useAgentChatHistory.
    public func refresh() async {
        guard let userId else { return }
        loadState = .loading
        do {
            let rows = try await AgentChatService.fetchThread(userId: userId, agentId: agentId)
            self.messages = rows
            self.loadState = .loaded
        } catch {
            self.loadState = .failed(Self.message(from: error))
        }
    }

    /// Send a user message + request an assistant reply. Optimistic on the user
    /// side; the assistant reply blocks `isAssistantTyping`.
    public func send() async {
        guard let userId, !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        let body = draft
        draft = ""
        let tempId = "temp-\(UUID().uuidString)"
        let temp = AgentChatMessage(
            id: tempId,
            avatarId: agentId,
            userId: userId,
            role: .user,
            content: body,
            createdAt: Self.nowISO()
        )
        messages.append(temp)

        do {
            let persisted = try await AgentChatService.sendUserMessage(
                userId: userId,
                agentId: agentId,
                content: body
            )
            // Swap temp with persisted row so we have the canonical id.
            if let idx = messages.firstIndex(where: { $0.id == tempId }) {
                messages[idx] = persisted
            }
        } catch {
            // Strip the temp on failure and surface the error.
            messages.removeAll { $0.id == tempId }
            lastError = Self.message(from: error)
            return
        }

        isAssistantTyping = true
        defer { isAssistantTyping = false }
        do {
            let reply = try await AgentChatService.requestAssistantReply(agentId: agentId)
            messages.append(reply)
        } catch {
            lastError = Self.message(from: error)
        }
    }

    // MARK: - Helpers

    private static func message(from error: Error) -> String {
        let raw = (error as NSError).localizedDescription
        return raw.isEmpty ? "Unknown error" : raw
    }

    private static func nowISO() -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f.string(from: Date())
    }
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/AgentChatStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.models.AgentChatMessage
import com.wagerproof.core.services.AgentChatService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/**
 * Port of iOS `AgentChatStore.swift` (doc §8.4). Per-agent chat thread (NOT
 * WagerBot). Optimistic user send: append a temp message (id = "temp-{uuid}"),
 * swap for the persisted row on success, strip on failure; then request the
 * assistant reply while [isAssistantTyping] is true (non-streaming — one awaited
 * reply message).
 *
 * Created per-screen: `AgentChatStore(agentId, userId)`.
 */
@Stable
class AgentChatStore(
    agentId: String,
    userId: String? = null,
    private val service: AgentChatService = AgentChatService,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    var agentId by mutableStateOf(agentId); private set
    var userId by mutableStateOf(userId); private set

    var messages by mutableStateOf<List<AgentChatMessage>>(emptyList()); private set
    var loadState by mutableStateOf<LoadState>(LoadState.Idle); private set
    var isAssistantTyping by mutableStateOf(false); private set
    var lastError by mutableStateOf<String?>(null); private set

    var draft by mutableStateOf("")

    fun bind(userId: String?) {
        if (userId == this.userId) return
        this.userId = userId
        messages = emptyList()
        loadState = LoadState.Idle
    }

    /** Load the full thread. Mirrors useAgentChatHistory. */
    suspend fun refresh() {
        val uid = userId ?: return
        loadState = LoadState.Loading
        runCatching { service.fetchThread(userId = uid, agentId = agentId) }
            .onSuccess { rows ->
                messages = rows
                loadState = LoadState.Loaded
            }
            .onFailure { loadState = LoadState.Failed(message(it)) }
    }

    /**
     * Send a user message + request an assistant reply. Optimistic on the user
     * side; the assistant reply blocks [isAssistantTyping].
     */
    suspend fun send() {
        val uid = userId ?: return
        if (draft.trim().isEmpty()) return
        val body = draft
        draft = ""
        val tempId = "temp-${java.util.UUID.randomUUID()}"
        val temp = AgentChatMessage(
            id = tempId,
            avatarId = agentId,
            userId = uid,
            role = AgentChatMessage.Role.USER,
            content = body,
            createdAt = nowISO(),
        )
        messages = messages + temp

        val persisted = runCatching {
            service.sendUserMessage(userId = uid, agentId = agentId, content = body)
        }.getOrElse {
            // Strip the temp on failure and surface the error.
            messages = messages.filterNot { m -> m.id == tempId }
            lastError = message(it)
            return
        }
        // Swap temp with persisted row so we have the canonical id.
        messages = messages.map { m -> if (m.id == tempId) persisted else m }

        isAssistantTyping = true
        try {
            runCatching { service.requestAssistantReply(agentId = agentId) }
                .onSuccess { reply -> messages = messages + reply }
                .onFailure { lastError = message(it) }
        } finally {
            isAssistantTyping = false
        }
    }

    fun close() = scope.cancel()

    // MARK: - Helpers

    private fun message(error: Throwable): String {
        val raw = error.message.orEmpty()
        return raw.ifEmpty { "Unknown error" }
    }

    private fun nowISO(): String =
        ISO_MILLIS.format(Instant.now())

    private companion object {
        // ISO8601 with fractional seconds + Z, matching Swift's
        // [.withInternetDateTime, .withFractionalSeconds].
        val ISO_MILLIS: DateTimeFormatter =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC)
    }
}
